### PR TITLE
third party oauth endpoints

### LINKED ISSUE #29: Create endpoints for OAuth
- [ ] Endpoints to store OAuth tokens
- [ ] Endpoints to check if an authenticated user has linked the accounts - (Github, Travis, Heroku)
- [ ] Create endor app on Github and Heroku - store client secret in endor
- [ ] Callback endpoints to provide to respective OAuth providers to post tokens

### PR DESCRIPTION
**Pull request by jackcmeyer**
closes #29 
* adds support for checking, adding, updating github authentication tokens for a user
* adds support for checking, adding, updating travis access tokens for a user
* adds support for the databases
* adds encryption of tokens

### CODE REVIEW

File: src/services/githubauth.service.js [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+   * githubAuthentications table. If it doesn't exist, will return false. Otherwise, it will
+   * check if the token in the table is valid on Github by making a request. If it's invalid, we
+   * will delete the github token. If it is valid, return true.
+   * @param userId the user id to check
+   * @returns {Boolean} true, if the token is valid, false otherwise.
+   */
+  async checkIfUserIsAuthenticatedOnGithub(userId) {
+    const token = await this.getGithubTokenForUser(userId);
+
+    if (token && await githubService.isUserAuthenticated(encryptionUtil.decrypt(token.token))) {
```
Comment (nischayv) [inline src/services/githubauth.service.js]: Don't await in the if condition. Do it before.

File: src/db/sequelize.js [hunk 2]
```diff
@@ -211,6 +211,26 @@ function initManually(database, username, password, options) {
...
+  const GithubTokenOwner = model.define('githubTokenOwner', {
+    // ASSOCIATIONS DEFINED BELOW
+    id: {
+      primaryKey: true,
+      type: UUID,
+      defaultValue: UUIDV4,
+    }
+  });
+  GithubToken.belongsToMany(User, { as: 'owner', through: 'githubTokenOwner' });
+  User.belongsToMany(GithubToken, { as: 'githubToken', through: 'githubTokenOwner' });
```
Comment (nischayv) [inline src/db/sequelize.js]: Shouldn't this be a one-one relationship rather than a many-many??

File: src/utils/encryption.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import crypto from 'crypto';
```
Comment (nischayv) [inline src/utils/encryption.js]: Why not use bcrypt instead of native crypto? It will be a little slower but much more secure.

Comment (jackcmeyer) [reply src/utils/encryption.js]: Bcrypt has no decrypt functionality

File: src/db/sequelize.js (hunk 2 above)
Comment (jackcmeyer) [reply src/db/sequelize.js]: Yes it should. But I couldn’t get it to work as sequelize would always give me a foreign key constraint error

File: src/services/githubauth.service.js (hunk 1 above)
Comment (jackcmeyer) [reply src/services/githubauth.service.js]: 👍

File: src/utils/encryption.js (hunk 3 above)
Comment (nischayv) [reply src/utils/encryption.js]: oh yeah nvm. Bcrypt hashes it, does not encrypt it.

File: src/db/sequelize.js (hunk 2 above)
Comment (nischayv) [reply src/db/sequelize.js]: Did you try adding the foreign keys in the models?

Comment (jackcmeyer) [reply src/db/sequelize.js]: Yup 🙁

**Pull request merged**